### ISSUE TITLE
[JST-14] Auto-batch write queries

### ISSUE THREAD

**Issue by dtbuchholz** [tablelandnetwork/js-tableland#138]
**Is your feature request related to a problem? Please describe.**
I would like to use `write` and simply insert _all_ of the values that I have (e.g., [this](https://github.com/duneanalytics/spellbook/blob/main/models/addresses/cex_addresses.sql) or something larger), but Tableland imposes a 35kb query limit. This means that I have to write my own methods for reducing some set of query values into 35kb groups, and then calling `write` multiple times with each of these groups.

**Describe the solution you'd like**
1. Create a new method that takes the table's name and the value tuples. Under the hood, this will enable the query chunking and also perform the set of writes:
`writeBatch('table_1_1', [(1, 'value_01') , (2, 'value_02')])`
2. Alternatively, maybe `write` could perform this instead. Assumes multiple statements are delimited by a `;` -- chunk these statements into 35kb queries and perform writes one after another. Or, assume there is just one statement like `INSERT INTO table VALUES (1, 'value_01') , (2, 'value_02')` -- break the VALUES tuples up into ~35kb groups, then perform the writes:
`write(statement, { mode: 'batchInserts' })` or `write(statement, { mode: 'batchValues' })`

**Describe alternatives you've considered**
Implementing the batching logic on my own.

**Additional context**
This would make it easier to write queries and think less about limits at the client level; lower onboarding friction.

<sub>[JST-14](https://linear.app/tableland/issue/JST-14/auto-batch-write-queries)</sub>

**Comment by joewagner** [comment]
NRE, flagging a related discussion https://github.com/orgs/tablelandnetwork/discussions/230

**Comment by dtbuchholz** [comment]
@joewagner Not sure if this is still on the radar (that discussion no longer exists), but I think it's a definite need—unless I'm doing something incorrectly.

I'm running into a `Error: BATCH_ERROR: statement size error: larger than specified max` error. It'd be nice if I could just pass a large set of queries to the SDK, and it just figures it out for me, instead of me needing to check for 35kb and then chunking the queries myself.

For example, I tried creating a statement like `INSERT INTO table_1_20(permit_id,race_team,is_holo,driver_id) VALUES(0,'KILL SWITCHES',0,0),(1,'KILL SWITCHES',0,1),...,(2887,'KILL SWITCHES',0,2887)` that inserts ~3k values:
```js
// Create array of objects { permitId: number, raceTeam: string, isHolo: number, driverId: number }
const values = Array.from({ length: 2888 }, (_, k) => k).map((v) => {
  return { permitId: v, raceTeam: "KILL SWITCHES", isHolo: 0, driverId: v };
});

const formatValues = values
  .map((v) => `(${v.permitId},'${v.raceTeam}',${v.isHolo},${v.driverId})`)
  .join();
const statement = `INSERT INTO ${tableName}(permit_id,race_team,is_holo,driver_id) VALUES${formatValues}`;
const { meta: insert } = await db.prepare(statement).run();
```
And fwiw, I tried the batch method to see if it got around the issue (and as expected, it didn't):
```js
const statements = values.map(
  (v) =>
    `INSERT INTO ${tableName}(permit_id,race_team,is_holo,driver_id) VALUES(${v.permitId},'${v.raceTeam}',${v.isHolo},${v.driverId})`
);
const prepares = statements.map((v) => {
  return db.prepare(v);
});
const { meta: insert } = await db.batch(prepares);
```

Oh, and while I'm on the topic, do you know if there's an easy way to batch bind? e.g., instead of the `formatValues` in the first example, I wanted to just pass the `values` array, unpack it, and then use each object's values with named parameters, like:
```js
const statement = `INSERT INTO ${tableName}(permit_id,race_team,is_holo,driver_id) VALUES($permitId,$raceTeam,$isHolo,$driverId)`;
const { meta: insert } = await db.prepare(statement).bind(...values).run();
```
In other words, run a bunch of `bind`s to `($permitId,$raceTeam,$isHolo,$driverId)` that would produce `INSERT INTO table_1_20(permit_id,race_team,is_holo,driver_id) VALUES(0,'KILL SWITCHES',0,0),(1,'KILL SWITCHES',0,1),...,(2887,'KILL SWITCHES',0,2887)`.

**Comment by joewagner** [comment]
@dtbuchholz I'm a little worried if we enable a single SDK function call to do this kind of thing, it could introduce some potentially complicated and confusing behavior/outcomes. 

A few examples of potential speed bumps:
- I found that on Ethereum mainnet the theoretical query size max is hit when the single block gas limit is exceeded, which can happen with a query a lot smaller than 35kb. discussed [here](https://github.com/tablelandnetwork/evm-tableland/issues/260)
- In a browser this would end up needed a bunch of wallet prompts that trigger transactions that span multiple blocks.
- If there are multiple transactions that span multiple blocks, how long to wait to consider a block confirmed before proceeding with the set of transactions?

Do you have a use-case we could try to work through?  I'm wondering if for realistic uses, the developer should have to write their own script(s) that handle the logic of dealing with this kind of complexity?

**Comment by dtbuchholz** [comment]
@joewagner gotcha. Maybe it's a docs thing that shows how to chunk queries up. Fwiw, this is related to a real use case described here (it's not an explicit issue brought up by the devs but something I stumbled upon while ideating with them): https://www.notion.so/textile/DR-VRS-concept-feedback-8e5da3013b8b4b709aa12c93bfa931a8

I get that it's not a perfect science due to block size, so is there a recommended way to bulk insert, say, 3k rows of data? In this example, I set some arbitrary chunking size to create a query with a safe-enough size: https://gist.github.com/dtbuchholz/15fcb612f4bf1b9ff24302d050a5e3ee

This implementation is a bit naive but I think demonstrates what they'll have to do. I think the general idea is something that will come up more in the future as teams migrate data (e.g., full collection metadata stored on IPFS) to Tableland. But again, I get that it might be more of a docs thing / put onto the dev to implement, so I want to know what I should show. 

(Fwiw, when I saw the `bulk` method, my initial reaction was that it should handle this use case.)

**Comment by joewagner** [comment]
@dtbuchholz I like the idea of adding to the docs with some suggestions for chunking queries that are larger than our limit, and some nuances of the problems that might arise.

As far as how we might help devs, I tinkered with some abstracted helpers that might be useful.  
Here's a gist with some of the stuff I came up with: https://gist.github.com/joewagner/f240b91246081dbd1202b13a0714f802

It's only for inserts, but I we could in theory come up with something similar for updates (and even deletes).
You can see there's some assumptions and opinionated choices being made.  I wouldn't be against exporting a cleaned up version of `helpBatchBulkInsert` (needs a better name) from the SDK's helpers if you think people would want it.
